### DISCUSSION TITLE
JAXB non-reactive and reactive handle XML namespace prefixes differently

### DISCUSSION THREAD

**Discussion by gnieser** [quarkusio/quarkus#31882] [Q&A]
Hello,

I am experiencing an unexpected difference in the way quarkus-resteasy-reactive-jaxb handles XML namespace prefixes when switching from non-reactive (io.quarkus:quarkus-resteasy-jaxb:2.16.4.Final) to reactive (io.quarkus:quarkus-resteasy-reactive-jaxb:2.16.4.Final).

**Reproducer**
The code defines (available at https://github.com/gnieser/quarkus-jaxb):
* A XML 'commons' namespace with prefix 'c' for a package org.acme.xml.commons.
  * Which contains an XML type. The NameType with a firstname and a lastname fields.
```java
@javax.xml.bind.annotation.XmlSchema(namespace = "commons.xml.acme.org", elementFormDefault = javax.xml.bind.annotation.XmlNsForm.QUALIFIED, xmlns = {
    @javax.xml.bind.annotation.XmlNs(namespaceURI = "commons.xml.acme.org", prefix = "c")
})
package org.acme.xml.commons;
``` 

The code also defines two additional namespaces that are exclusive to each other and never use in the same XML:
* A XML 'acme' namespace with prefix 'acme' for a package org.acme.xml.persons
  * Which contains a list of Persons. It's a list of objects of type NameType.
```java
@javax.xml.bind.annotation.XmlSchema(namespace = "persons.xml.acme.org", elementFormDefault = javax.xml.bind.annotation.XmlNsForm.QUALIFIED, xmlns = {
    @javax.xml.bind.annotation.XmlNs(namespaceURI = "persons.xml.acme.org", prefix = "acme")
})
package org.acme.xml.persons;
```

* A XML 'acme' namespace with prefix 'acme' for a package org.acme.xml.users
  * Which contains a list of Users. It's a list of objects of type NameType. 
```java
@javax.xml.bind.annotation.XmlSchema(namespace = "users.xml.acme.org", elementFormDefault = javax.xml.bind.annotation.XmlNsForm.QUALIFIED, xmlns = {
    @javax.xml.bind.annotation.XmlNs(namespaceURI = "users.xml.acme.org", prefix = "acme")
})
package org.acme.xml.users;
```

**Issue**
The troublesome element is that while these two XML namespaces are never used together in the same response they define the same prefix for their respective namespace.
It may look weird, but this is a simplification of what a set of third party XSD schemas would produce through XJC. I have simplified to extreme and removed XJC from the equation.

**Using quarkus-resteasy-jaxb**
Only the namespaces in use in the Response are added to the XML so everything works fine:
```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<acme:users xmlns:acme="users.xml.acme.org" xmlns:c="commons.xml.acme.org">
  <acme:user>
    <c:firstName>Rosalind</c:firstName>
    <c:lastName>Franklin</c:lastName>
  </acme:user>
</acme:users>
```

**Using quarkus-resteasy-reactive-jaxb**
Even the namespaces not in use are added to the Response. So ensues a conflict in the prefixes, and one of them is set to ns2 and the XML response becomes:
```xml
<ns2:users xmlns:acme="persons.xml.acme.org" xmlns:ns2="users.xml.acme.org" xmlns:c="commons.xml.acme.org">
  <ns2:user>
    <c:firstName>Rosalind</c:firstName>
    <c:lastName>Franklin</c:lastName>
  </ns2:user>
</ns2:users>
```

**Question**
I understand that part of the issue is the conflicting prefixes, but I feel like something may be done to prevent:
* an unnecessary namespace definition from being added to the XML
* the behavior change between non-reactive to reactive

How can I guide quarkus-resteasy-reactive-jaxb to only add XML namespace definitions for the namespace that are actually part of the Response, and behave more like its non-reactive counterpart?

Note that even with a change to the prefixes (set to 'u' and 'p'), the unused namespace is still added to the XML:
```xml
<u:users xmlns:p="persons.xml.acme.org" xmlns:u="users.xml.acme.org" xmlns:c="commons.xml.acme.org">
  <u:user>
    <c:firstName>Rosalind</c:firstName>
    <c:lastName>Franklin</c:lastName>
  </u:user>
</u:users>
```

Thanks!

[a comment by an automated account is not shown]

**Reply by Sgitario** [reply]
This was happening because, for RESTEasy Reactive, we were binding all the discovered XML types into the same JAXBContext to get some performance benefits. 

In RESTEasy Classic, one JAXBContext by XML type was created.

This was fixed in Quarkus 3.0.0.Alpha5, so if you use this version, you will see the same behavior between RESTEasy Reactive and Classic (I did confirm it using your repository).
Additionally, you can now control the XML types to be bound into the default JAXBContext by excluding classes using the property `quarkus.jaxb.exclude-classes`). See more at https://quarkus.io/version/main/guides/resteasy-reactive#xml-serialisation.

**Reply by gnieser** [reply]
Thanks a lot for your answer! 

I get the expected behavior using:
* Quarkus 3.0.0.Alpha5
* and adding
```
quarkus.jaxb.exclude-classes=\
  org.acme.xml.persons.ObjectFactory,\
  org.acme.xml.persons.Persons,\
  org.acme.xml.users.ObjectFactory,\
  org.acme.xml.users.Users
```

May I ask whether there's a way to exclude by package / wildcards?
The motivation behind the question goes back to the scenario that let me to this reproducer. I'm using third-party XSD with XJC, and there are many generated classes.

**Reply by Sgitario** [reply]
That's a very fair request and makes total sense to me. 
The logic that excludes classes is located [here](https://github.com/quarkusio/quarkus/blob/45e4349ae506f884e5b5d1796e033da3cb2226df/extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/FilteredJaxbClassesToBeBoundBuildItem.java#L54). Would you like to contribute with this enhancement? :)

**Reply by gnieser** [reply]
I would be honored to contribute, if I'm able to. I'm reading the guide, trying to compile a forked main branch (IDE seems pretty unresponsive though). 

What would be the ideal solution?
* A quarkus.jaxb.exclude-packages. It excludes all classes within the packages (and sub-packages)
* A wildcard approach of some sort, with Java Regex or Ant-style (such as https://quarkus.io/guides/all-config#quarkus-arc_quarkus.arc.exclude-types or https://quarkus.io/guides/all-config#quarkus-core_quarkus.test.exclude-pattern) ? If so I suppose there may be an utility class I can leverage within the existing code base.

**Reply by gnieser** [reply]
I went for both. :)
Would [this](https://github.com/quarkusio/quarkus/compare/main...gnieser:quarkus:jaxbcontext-excludes) be suitable?

**Reply by Sgitario** [reply]
These changes sound really good! Many thanks! 
Though, I would merge the `exclude-classes` and `exclude-packages` into `exclude-classes` (and use the `startWith` matching). 
Moreover, I would not add pattern matching at least it's essential since we used not to try adding too many superficial configuration options.
But this is something we can discuss in a pull request, so we could incorporate more people's opinions.

**Reply by gnieser** [reply]
As per your suggestion, I have opened https://github.com/quarkusio/quarkus/pull/31940
This is my first PR, hopefully I'm not doing anything wrong with the process.
